pyuvdata/uvdata.py

```diff
@@ -4628,3 +4654,497 @@ def inflate_by_redundancy(self, tol=1.0, blt_order='time', blt_minor_order=None)
+                    temp_time[temp_idx] = np.average(self.time_array[averaging_idx])
+                    temp_int_time[temp_idx] = running_int_time
+                    if not self.metadata_only:
+                        # if all inputs are flagged, the flag array should be True,
+                        # otherwise it should be False.
+                        # The sum below will be zero if it's all flagged and
+                        # greater than zero otherwise
+                        # Then we use a test against 0 to turn it into a Boolean
+                        temp_flag[temp_idx] = np.sum(~self.flag_array[averaging_idx], axis=0) == 0
+
+                        mask = self.flag_array[averaging_idx]
+                        # need to update mask if a downsampled visibility will be flagged
+                        # so that we don't set it to zero
+                        if (temp_flag[temp_idx]).any():
+                            ax1_inds, ax2_inds, ax3_inds = np.nonzero(temp_flag[temp_idx])
+                            for ii in range(ax1_inds.size):
+                                mask[:, ax1_inds[ii], ax2_inds[ii], ax3_inds[ii]] = False
+
+                        masked_data = np.ma.masked_array(self.data_array[averaging_idx],
+                                                         mask=mask)
+                        if summing_correlator_mode:
+                            temp_data[temp_idx] = np.sum(masked_data, axis=0)
+                        else:
+                            temp_data[temp_idx] = np.mean(masked_data, axis=0)
+                        # nsample array is the fraction of data that we actually kept,
+                        # relative to the amount that went into the sum or average
+                        masked_nsample = np.ma.masked_array(self.nsample_array[averaging_idx],
+                                                            mask=mask)
+                        temp_nsample[temp_idx] = (np.sum(masked_nsample, axis=0)
+                                                  / float(self.flag_array[averaging_idx].shape[0]))
+                    # increment counters and reset values
+                    temp_idx += 1
+                    summing_idx += n_sum
+                    running_int_time = 0.0
+                    n_sum = 0
+
+        # make sure we've populated the right number of baseline-times
+        assert temp_idx == temp_Nblts, ("Wrong number of baselines. Got {:d}, "
+                                        "expected {:d}. This is a bug, please "
+                                        "make an issue at https://github.com/"
+                                        "RadioAstronomySoftwareGroup/pyuvdata/"
+                                        "issues".format(temp_idx, temp_Nblts))
+
+        # drop data where we downsampled
+        inds_to_keep = np.nonzero(self.integration_time >= min_int_time)
+        self.baseline_array = self.baseline_array[inds_to_keep]
```

Comment:
Basically from here to the end of the function is repeated between up/down sampling. Would it make sense to abstract this out to a function so we don't have so much repeated code?

Reply:
Yeah, probably better to have less repeated code. I've pulled out most of the repeated bits into a common function.